### PR TITLE
New package: freecad-0.16

### PR DESCRIPTION
**Pull request by yopito**
runtime is fine with x86_64/glibc.
cross-build not yet tested, let's see what's Travis CI think about it.
might also rename some packages by prefixing them with "coin3-" ?

### CODE REVIEW

**Comment by yopito** [comment]
rebased: fix musl (native) build for coin3 and oce packages.

Note that Travis CI is not that helpfull: some build stops because log file is 4MB, other because build time exceeded 120 minutes.

Also has troubles with cross-building, since coin3 provides a shell script `coin3-config`, so it should be in `hostmakedepend` ... might find a workaround if relying on `pkgconfig` stuff instead ?

Any help is welcome ;)

File: common/shlibs [hunk 1]
```diff
@@ -2458,3 +2458,67 @@ libmarkdown.so.2 discount-2.2.0_3
...
+libTKXDESTEP.so.10 oce-0.17.1_1
+libTKXMesh.so.10 oce-0.17.1_1
+libTKXSBase.so.10 oce-0.17.1_1
+libTKXSDRAW.so.10 oce-0.17.1_1
+libTKXml.so.10 oce-0.17.1_1
+libTKXmlL.so.10 oce-0.17.1_1
+libTKXmlTObj.so.10 oce-0.17.1_1
+libTKXmlXCAF.so.10 oce-0.17.1_1
+libTKernel.so.10 oce-0.17.1_1
+libCoin.so.60 coin3-3.1.3_1
```
Comment (ebfe) [inline common/shlibs]: these probably belong into the coin3 commit

File: srcpkgs/gl2ps/template [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+maintainer="yopito <pierre.bourgin@free.fr>"
+license="LGPL-2, GL2PS"
+homepage="http://www.geuz.org/gl2ps/"
+distfiles="http://geuz.org/${pkgname}/src/${pkgname}-${version}.tgz"
+checksum=8a680bff120df8bcd78afac276cdc38041fed617f2721bade01213362bcc3640
+
+post_install() {
+	rm ${DESTDIR}/usr/lib/*.a
+	vlicense COPYING.GL2PS
+	vcopy gl2ps.pdf    usr/share/doc/${pkgname}
```
Comment (ebfe) [inline srcpkgs/gl2ps/template]: vdoc gl2ps.pdf

Comment (ebfe) [inline srcpkgs/gl2ps/template]: and maybe move to the -devel package?

File: srcpkgs/coin3/template [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+patch_args="-Np1"
+build_style=gnu-configure
+configure_args="--includedir=/usr/include/Coin3 --disable-dependency-tracking
+ --disable-static --enable-threadsafe --enable-exceptions
+ --disable-dl-simage --disable-dl-glu --disable-dl-libbzip2
+ --disable-dl-zlib --disable-dl-freetype --disable-dl-fontconfig
+ --enable-3ds-import --enable-html"
+hostmakedepends="pkg-config doxygen perl"
+makedepends="freetype-devel glu-devel zlib-devel bzip2-devel fontconfig-devel
+ freetype-devel boost-devel simage-devel"
```
Comment (Vaelatern) [inline srcpkgs/coin3/template]: `freetype-devel` is repeated here

File: srcpkgs/freecad/template [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+license="GPL-2"
+homepage="http://freecadweb.org/"
+distfiles="https://github.com/FreeCAD/FreeCAD/archive/${version}.tar.gz"
+checksum=6cc71ab4b0dc60b493d3aaa4b42f1ce1af9d4fcd539309ab0792804579e18e09
+
+post_install() {
+
+	# AppHomePath is computed with binary's realpath:
+	# do not move binaries but symlink them instead.
+	mkdir ${DESTDIR}/usr/bin
```
Comment (Vaelatern) [inline srcpkgs/freecad/template]: `vmkdir usr/bin`

File: srcpkgs/freecad/template [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+	for f in FreeCAD FreeCADCmd; do
+		ln -s ${_inst_prefix}/bin/${f} ${DESTDIR}/usr/bin/${f}
+	done
+
+	# desktop integration
+	mkdir -p ${DESTDIR}/usr/share/applications
```
Comment (Vaelatern) [inline srcpkgs/freecad/template]: `vmkdir usr/share/applications`

File: srcpkgs/freecad/template [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+	vcopy ${FILESDIR}/freecad.appdata.xml usr/share/appdata
+	mkdir -p ${DESTDIR}/usr/share/mime/packages
+	vcopy ${FILESDIR}/freecad.xml usr/share/mime/packages
+	for s in 16 32 48 64; do
+		_dest_icon=/usr/share/icons/hicolor/${s}x${s}/apps
+		mkdir -p ${DESTDIR}/${_dest_icon}
+		ln -s ${_inst_prefix}/data/freecad-icon-${s}.png ${DESTDIR}/${_dest_icon}/freecad.png
+		ln -s ${_inst_prefix}/data/freecad.xpm ${DESTDIR}/${_dest_icon}/freecad.xpm
+	done
+	mkdir -p ${DESTDIR}/usr/share/icons/scalable/apps
```
Comment (Vaelatern) [inline srcpkgs/freecad/template]: `vmkdir usr/share/icons/scalable/apps`

File: srcpkgs/freecad/template [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+		ln -s ${_inst_prefix}/bin/${f} ${DESTDIR}/usr/bin/${f}
+	done
+
+	# desktop integration
+	mkdir -p ${DESTDIR}/usr/share/applications
+	vcopy ${FILESDIR}/freecad.desktop usr/share/applications
+	sed -i -e "s,@_inst_prefix@,${_inst_prefix}," ${DESTDIR}/usr/share/applications/freecad.desktop
+	mkdir -p ${DESTDIR}/usr/share/appdata
+	vcopy ${FILESDIR}/freecad.appdata.xml usr/share/appdata
+	mkdir -p ${DESTDIR}/usr/share/mime/packages
```
Comment (Vaelatern) [inline srcpkgs/freecad/template]: `vmkdir`

File: srcpkgs/freecad/template [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+	mkdir ${DESTDIR}/usr/bin
+	for f in FreeCAD FreeCADCmd; do
+		ln -s ${_inst_prefix}/bin/${f} ${DESTDIR}/usr/bin/${f}
+	done
+
+	# desktop integration
+	mkdir -p ${DESTDIR}/usr/share/applications
+	vcopy ${FILESDIR}/freecad.desktop usr/share/applications
+	sed -i -e "s,@_inst_prefix@,${_inst_prefix}," ${DESTDIR}/usr/share/applications/freecad.desktop
+	mkdir -p ${DESTDIR}/usr/share/appdata
```
Comment (Vaelatern) [inline srcpkgs/freecad/template]: `vmkdir`

File: srcpkgs/gl2ps/template (hunk 2 above)
Comment (Vaelatern) [inline srcpkgs/gl2ps/template]: Yeah, the `*.a` in particular should go in the `-devel` package

File: srcpkgs/libspnav/template [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+makedepends="libX11-devel"
+short_desc="Open source alternative to 3DConnextion drivers"
+maintainer="yopito <pierre.bourgin@free.fr>"
+license="GPL-3"
+homepage="http://spacenav.sourceforge.net/"
+distfiles="${SOURCEFORGE_SITE}/spacenav/${pkgname}-${version}.tar.gz"
+checksum=7ae4d7bb7f6a5dda28b487891e01accc856311440f582299760dace6ee5f1f93
+
+post_install() {
+	rm ${DESTDIR}/usr/lib/libspnav.a
```
Comment (Vaelatern) [inline srcpkgs/libspnav/template]: Instead, in the `-devel` package, `vmove usr/lib/*.a`

File: common/shlibs (hunk 1 above)
Comment (yopito) [inline common/shlibs]: my bad, thanks for pointing this.

File: srcpkgs/coin3/template [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+post_extract() {
+	# remove bundled boost headers
+	rm -rf ${wrksrc}/include/boost
+}
+
+do_configure() {
+	export CPPFLAGS+="$(pkg-config --cflags freetype2)"
+	: ${configure_script:=./configure}
+	${configure_script} ${configure_args}
+}
```
Comment (yopito) [inline srcpkgs/coin3/template]: by irc: test `export CPPFLAGS+="$(pkg-config --cflags freetype2)"` in the core of the template and remove `do_configure()` then

**Comment by yopito** [comment]
rebased to fix all comments.
FIY, x86_64/musl native builds succeeded, have to test its runtimes though.

**Comment by yopito** [comment]
Rebased: 
* small fixes
* flag packages that needs `nocross=yes` 
* added `pycompile_dirs`  directive into FreeCAD's template.

As usual, Travis CI will probably timeout because of build lenght (2+ hours) or log file (4+ MB).

I've tested the runtime with x86_64/glibc and x86_64/musl, including opening the provided samples and using the internal tests of the application (`FreeCADCmd -t 1`).

**Comment by yopito** [comment]
rebased again, sorry: fix cross-build of libspnav

**Comment by yopito** [comment]
rebased on top of master, no change.

**Comment by karl-nilsson** [comment]
* FreeCAD's license is LGPL2+
https://github.com/FreeCAD/FreeCAD/blob/releases/FreeCAD-0-16/copying.lib
* SoQt is not required, and its replacement (Quarter) is already part of the freecad source tree
http://forum.freecadweb.org/viewtopic.php?f=8&t=733&start=80#p117634

Additional Requirements (starting here: https://github.com/FreeCAD/FreeCAD/blob/releases/FreeCAD-0-16/CMakeLists.txt#L568

* zlib
* boost
* PyCXX

**Comment by yopito** [comment]
Commits just updated (squashed). See details below of what has changed.

@karl-nilsson : many thanks for your comments. Here my replies:

* FreeCAD license: fixed

* SoQt: was a dep of python-pivy package. Checking its source code shows that it uses Quarter if no other toolkit is present (SoQt, SoGTK, etc). So you're right ! Sorry for that (its build output did not notice that). So I removed the SoQt package and the corresponding dep in my python-pivy package

* FreeCAD / zlib : indirect dependency (not explicitely required), but it's present. FreeCAD cmake's output:  
    ```
...
-- Found Xerces-C: /usr/lib64/libxerces-c.so
-- Found ZLIB: /usr/lib64/libz.so (found version "1.2.8") 
    ```

* FreeCAD / boost: already present in `freecad/template`. Confirmed by cmake output:  
    ```
...
-- Boost version: 1.58.0
-- Found the following Boost libraries:
--   filesystem
--   program_options
--   regex
--   signals
--   system
--   thread
--   chrono
--   date_time
--   atomic
    ```

* FreeCAD / PyCXX: PyCXX is not (yet ?) packaged in XBPS repo, so FreeCAD uses its internal copy. Is its a problem ? cmake's output:
    ```
...
-- PyCXX found:
--   Headers:  /builddir/FreeCAD-0.16/src
--   Sources:  /builddir/FreeCAD-0.16/src/CXX
    ```

* FreeCAD and OCE/OCCT: please notice that I provide a packaging against OCE "only" ;)

**Comment by yopito** [comment]
updated (and squash): oce build is broken on arm. will try to fix it later

**Comment by yopito** [comment]
rebased, updated (and squashed): few updates on packaging comments (discussed with @Vaelatern on IRC).

@karl-nilsson reported (IRC) successfull package builds and basic runtime usage of FreeCAD is fine

**Comment by Vaelatern** [comment]
Closed.
https://git.io/vKtuS

**Pull request closed without merging**